kore/src/Kore/Rewrite/Axiom/Identifier.hs

```diff
@@ -74,8 +75,12 @@ data AxiomIdentifier
       Equals !AxiomIdentifier !AxiomIdentifier
     | -- | An @\\exists@ pattern with the given child.
       Exists !AxiomIdentifier
+    | -- | A @\\not@ pattern with the given child.
+      Not !AxiomIdentifier
     | -- | Any variable pattern.
       Variable
+    | -- | Anything else.
+      Other
```

Comment:
Regarding the discussion here: https://github.com/runtimeverification/haskell-backend/issues/3098#issuecomment-1156671650

Is adding this constructor any different to just removing the `AxiomIdentifier` arguments in this type?

My reasoning is that this function was partial so that would shortcircuit attempts to apply equations whenever we'd try to simplify terms which are unlikely to be simplified by rules provided by the user.

Reply:
> Is adding this constructor any different to just removing the `AxiomIdentifier` arguments in this type?

Adding this "catch-all" constructor means any (simplification) equation from the user will get added, under a non-specific `Other` key if nothing more specific matches.
If we removed the `AxiomIdentifier` recursion in the type instead, we would store only what we had stored before but under less specific keys. 
For illustration: If we remove the recursion, _any_ `ceil` simplification equations will be stored under a single `Ceil` key, and rules like (1) `\ceil (\dv ...) => ...` and (2) `\ceil (\not (\dv ...)) => ...` will both be tried on any term that has a `\ceil` at the top (say, (3) `\ceil (\and ...)`. 
In the proposed implementation, the keys would be (1) `Ceil  DV`, (2) `Ceil (Not DV)` and (3) `Ceil Other`, previously (3) would have no key, and rules with LHS like (3) would be dropped entirely.

We could maybe filter out the plain `Other` at the top (after the `fold` returns), that would retain some of the prior partiality. LMK if you would prefer that.

Reply:
👍